### PR TITLE
Conditionally mark getter/setter implementations virtual in generated classes

### LINKED ISSUE #8302: Generated get/set methods incorrectly marked non-virtual
While testing an experimental feature to [add support for interface declarations](https://github.com/PowerShell/PowerShell/issues/2223#issuecomment-439602845), I found that types generated using the `class` keyword can't correctly override get/set methods associated with properties.

# Steps to reproduce

```powershell
Add-Type @'
namespace TypeTest {
  public interface Interface {
    string S { get; set; }
  }
}
'@

# Execute this as a separate statement to ensure TypeTest.Interface is defined before the class definition is
class TestClass : TypeTest.Interface {
  [string]$S
}

```

# Expected behavior

New type `[TestClass]` implementing `[TypeTest.Interface]` should have been compiled and loaded.

# Actual behavior

```none
At line:1 char:1
+ class TestClass : TypeTest.Interface {
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Error during creation of type "TestClass". Error message:
Method 'get_S' in type 'TestClass' from assembly 'Powershell Class Assembly, Version=1.0.0.1, Culture=neutral, PublicKeyToken=null' does not have an implementation.
    + CategoryInfo          : ParserError: (:) [], ParentContainsErrorRecordException
    + FullyQualifiedErrorId : TypeCreationError
```

# Solution

After comparing the CIL emitted by the C# compiler when implicitly implementing an interface property, I found that the get/set methods generated by the compiler are marked `virtual`.

The fix is to either attempt to discover properties with matching signatures in associated instances and mark the backing get/set methods `virtual`, or simply just mark *all* get/set methods `virtual` (might incur slight overhead in method resolution at runtime rather than at compile time)

# Environment data

```none

Name                           Value
----                           -----
PSVersion                      6.2.0-preview.1
PSEdition                      Core
GitCommitId                    6.2.0-preview.1-79-g16569b157a430cc5d4a2e7bf189dcfef64beb31f
OS                             Microsoft Windows 10.0.17134 
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

### PR DESCRIPTION
**Pull request by IISResetMe**
Fix https://github.com/PowerShell/PowerShell/issues/8302

When implementing interfaces, PowerShell incorrectly produces non-virtual get/set methods for interface-defined properties.
This commit adds a lookup method for interface-defined properties and marks get/set methods for properties with matching signatures virtual.

## PR Checklist

- [X] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [X] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [X] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **User-facing changes**
    - [X] Not Applicable
    - **OR**
    - [ ] User-facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed - Issue link:
- **Testing - New and feature**
    - [] Not Applicable or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [ ] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

I'm unsure how to meaningfully add tests to prevent regression for this

### CODE REVIEW

**Comment by IISResetMe** [comment]
@iSazonov if you have anything that's either more correct, complete or aesthetically pleasing, please feel free to update the PR :)

**Comment by iSazonov** [comment]
@IISResetMe My post was informational. If you commit will approved I can grab it to my branch or rebase my branch.

**Comment by IISResetMe** [comment]
@iSazonov Fixed the CodeFactor issues in `ShouldImplementProperty()` but I'm unsure what to do with the "Complex Method" finding for `EmitPropertyIL()`

**Comment by iSazonov** [comment]
We ignore "Complex Method".

File: test/powershell/Language/Classes/scripting.Classes.inheritance.tests.ps1 [hunk 1]
```diff
@@ -63,6 +63,14 @@ Describe 'Classes inheritance syntax' -Tags "CI" {
         [MyComparable].GetInterface("System.IComparable") | Should -Not -BeNullOrEmpty
     }
 
+    It 'can implement .NET interface properties' {
+        Add-Type -TypeDefinition 'public interface I1 { int I{ get; set; } }'
```
Comment (iSazonov) [inline test/powershell/Language/Classes/scripting.Classes.inheritance.tests.ps1]: Please replace "I" with more readable name. Maybe `InterfaceProperty`?

File: test/powershell/Language/Classes/scripting.Classes.inheritance.tests.ps1 [hunk 2]
```diff
@@ -63,6 +63,14 @@ Describe 'Classes inheritance syntax' -Tags "CI" {
...
+        $C1 = Invoke-Expression 'class C1 : I1 {[int]$I}'
+        $getter = $C1.GetType().GetMember('get_I')
+        $getter.ReturnType |Should -BeOfType System.Int32
```
Comment (iSazonov) [inline test/powershell/Language/Classes/scripting.Classes.inheritance.tests.ps1]: Please add space before `Should`.

File: src/System.Management.Automation/engine/parser/PSType.cs [hunk 3]
```diff
@@ -434,6 +434,19 @@ private Type GetBaseTypes(Parser parser, TypeDefinitionAst typeDefinitionAst, ou
                 return baseClass ?? typeof(object);
             }
 
+            private bool ShouldImplementProperty(string name, Type type)
+            {
+                foreach (var @interface in _typeBuilder.GetInterfaces())
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/parser/PSType.cs]: I'd not use a reserved name. Maybe `interfaceType`?

File: test/powershell/Language/Classes/scripting.Classes.inheritance.tests.ps1 (hunk 1 above)
Comment (IISResetMe) [reply test/powershell/Language/Classes/scripting.Classes.inheritance.tests.ps1]: Update both type names and the property name to be more descriptive

**Review by iSazonov — APPROVED** [review]

**Review by lzybkr — APPROVED** [review]

**Comment by iSazonov** [comment]
Reopen to restart Appveyor CI.,

**Comment by iSazonov** [comment]
@IISResetMe Thanks for your contribution!

**Pull request merged**